clippy_lints/src/functions/result.rs

```diff
@@ -0,0 +1,152 @@
+use rustc_errors::Diagnostic;
+use rustc_hir as hir;
+use rustc_lint::{LateContext, LintContext};
+use rustc_middle::lint::in_external_macro;
+use rustc_middle::ty;
+use rustc_span::{sym, Span};
+use rustc_typeck::hir_ty_to_ty;
+
+use clippy_utils::diagnostics::{span_lint_and_help, span_lint_and_then};
+use clippy_utils::trait_ref_of_method;
+use clippy_utils::ty::is_type_diagnostic_item;
+
+use super::{RESULT_LARGE_ERR, RESULT_UNIT_ERR};
+
+/// comes up with an "at least" guesstimate for the type's size, not taking into
+/// account the layout of type parameters.
+/// For example, a `((u32, u32, T), T, [u8; 4])` will be 12, which is only
+/// correct if `T` is a zst *and* doesn't have alignment...
+fn approx_ty_size<'tcx>(cx: &LateContext<'tcx>, ty: ty::Ty<'tcx>) -> u64 {
+    use rustc_middle::ty::layout::LayoutOf;
+    match (cx.layout_of(ty).map(|layout| layout.size.bytes()), ty.kind()) {
+        (Ok(size), _) => size,
```

Comment:
Could include arrays with something like

```suggestion
        (Ok(size), _) => size,
        (Err(_), ty::Array(t, n)) => n.try_eval_usize(cx.tcx, cx.param_env).map_or(0, |len| len * approx_ty_size(cx, t))
```

Reply:
👍 I didn't think for `[(T, u128); 10]` et al